src/openrct2/core/Http.WinHttp.cpp

```diff
@@ -0,0 +1,223 @@
+                return L"GET";
+            case Method::POST:
+                return L"POST";
+            case Method::PUT:
+                return L"PUT";
+            default:
+                throw std::runtime_error("Unsupported verb.");
+        }
+    }
+
+    static std::wstring ReadHeader(HINTERNET hRequest, DWORD infoLevel, const wchar_t* name)
+    {
+        wchar_t headerBuffer[256]{};
+        auto headerBufferLen = (DWORD)std::size(headerBuffer);
+        if (!WinHttpQueryHeaders(hRequest, infoLevel, name, headerBuffer, &headerBufferLen, WINHTTP_NO_HEADER_INDEX))
+            ThrowWin32Exception("WinHttpQueryHeaders");
+        return std::wstring(headerBuffer, headerBufferLen);
+    }
+
+    static int32_t ReadStatusCode(HINTERNET hRequest)
+    {
+        auto wStatusCode = ReadHeader(hRequest, WINHTTP_QUERY_STATUS_CODE, L"StatusCode");
+        return std::stoi(wStatusCode);
+    }
+
+    static std::map<std::string, std::string> ReadHeaders(HINTERNET hRequest)
+    {
+        DWORD dwSize{};
+        WinHttpQueryHeaders(
+            hRequest, WINHTTP_QUERY_RAW_HEADERS, WINHTTP_HEADER_NAME_BY_INDEX, NULL, &dwSize, WINHTTP_NO_HEADER_INDEX);
+
+        std::wstring buffer;
+        buffer.resize(dwSize);
+        if (!WinHttpQueryHeaders(
+                hRequest, WINHTTP_QUERY_RAW_HEADERS, WINHTTP_HEADER_NAME_BY_INDEX, buffer.data(), &dwSize,
+                WINHTTP_NO_HEADER_INDEX))
+            ThrowWin32Exception("WinHttpQueryHeaders");
+
+        std::map<std::string, std::string> headers;
+        std::wstring wKey, wValue;
+        int state = 0;
+        int index = 0;
+        for (auto c : buffer)
+        {
+            if (c == '\0')
+            {
+                state = 0;
+                if (index != 0 && wKey.size() != 0)
+                {
+                    auto key = String::ToUtf8(wKey);
+                    auto value = String::ToUtf8(wValue);
+                    headers[key] = value;
+                }
+                wKey.clear();
+                wValue.clear();
+                index++;
+                continue;
+            }
+            if (state == 0 && c == ':')
+            {
+                state = 1;
+            }
+            else if (state == 1 && c == ' ')
+            {
+                state = 2;
+            }
+            else if (state == 0)
+            {
+                wKey.push_back(c);
+            }
+            else
+            {
+                state = 2;
+                wValue.push_back(c);
+            }
+        }
+        return headers;
+    }
+
+    static std::string ReadBody(HINTERNET hRequest)
+    {
+        DWORD dwSize{};
+        std::string body;
+        do
+        {
+            // Check for available data.
+            if (!WinHttpQueryDataAvailable(hRequest, &dwSize))
+                ThrowWin32Exception("WinHttpQueryDataAvailable");
+
+            // No more available data.
+            if (dwSize == 0)
+                break;
```

Comment:
Remarks from MSDN:
```
Important  Do not use the return value of WinHttpQueryDataAvailable to determine whether the end of a response has been reached, because not all servers terminate responses properly, and an improperly terminated response causes WinHttpQueryDataAvailable to anticipate more data.
```
They explicitly say
```
Use the return value of WinHttpReadData to determine when a response has been completely read.
```
See https://docs.microsoft.com/en-us/windows/win32/api/winhttp/nf-winhttp-winhttpquerydataavailable

Reply:
Hmm, this snippet of code was copied from their example.

Reply:
The documentation says do not use it, dunno if the example is bad or not but I would rather trust the documentation over snippets made up by interns. It has probably to do with servers that do not transfer the Content-Length field and relies on terminating the connection when the response completes.

Reply:
Can you suggest the change to make?

Reply:
According to MSDN:
```If you are using WinHttpReadData synchronously, and the return value is TRUE and the number of bytes read is zero, the transfer has been completed and there are no more bytes to read on the handle. This is analogous to reaching end-of-file in a local file.```